### ISSUE TITLE
Potential CSP violations

### ISSUE THREAD

**Issue by rodneyrehm** [medialize/ally.js#129]
The [CSP (Content Security Policy)](https://developer.mozilla.org/en-US/docs/Web/Security/CSP) can be used on any document to limit the document's scripting ability. [caniuse](http://caniuse.com/#search=CSP) says Chrome and Firefox support this feature already.

The supports tests do things like creating `<object src="data:…">` to determine if they're focusable or not and these tests may violate a CSP.

---

This problem was first spotted by @ryan-ludwig with [Chrome Apps](https://developer.chrome.com/apps/contentSecurityPolicy), which sport the following (not relaxable) CSP:

```
Content-Security-Policy:
  default-src 'self';
  connect-src * data: blob: filesystem:;
  style-src 'self' data: chrome-extension-resource: 'unsafe-inline';
  img-src 'self' data: chrome-extension-resource:;
  frame-src 'self' data: chrome-extension-resource:;
  font-src 'self' data: chrome-extension-resource:;
  media-src * data: blob: filesystem:;
```

The error they observed was:

> "Refused to load plugin data from 'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zd…V4dCB4PSIxMCIgeT0iMjAiIGlkPSJzdmctbGluay10ZXh0Ij50ZXh0PC90ZXh0Pjwvc3ZnPg==' because it violates the following Content Security Policy directive: "default-src 'self' blob: filesystem: chrome-extension-resource:". Note that 'object-src' was not explicitly set, so 'default-src' is used as a fallback."

The problem in this particular case is that the directive `object-src 'self' data:;` is missing. See the potential CSP violations below.

## CSP directives currently required by ally.js 1.1.0

After running the test suite with [`Content-Security-Policy-Report-Only`](https://developer.mozilla.org/en-US/docs/Web/Security/CSP/Using_CSP_violation_reports) the following violations were logged, forming this policy:

```
Content-Security-Policy:
  default-src 'self';
  style-src 'unsafe-inline';
  img-src data:;
  media-src data:;
  object-src 'self' data:;
  frame-src data:;
```

### style-src: inline

covered by directive `style-src 'unsafe-inline';`

* [supports/detect-focus.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/detect-focus.js#L15)
* [supports/focus-children-of-focusable-flexbox.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-children-of-focusable-flexbox.js#L11-L12)
* [supports/focus-flexbox-container.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-flexbox-container.js#L10-L11)
* [supports/focus-scroll-body.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-scroll-body.js#L10-L11)
* [supports/focus-scroll-container-without-overflow.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-scroll-container-without-overflow.js#L10-L11)
* [supports/focus-scroll-container.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-scroll-container.js#L10-L11)

### img-src: data

covered by directive `img-src data:;`

* [supports/focus-area-img-tabindex.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-area-img-tabindex.js#L13)
* [supports/focus-area-tabindex.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-area-tabindex.js#L14)
* [supports/focus-area-without-href.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-area-without-href.js#L14)
* [supports/focus-broken-image-map.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-broken-image-map.js#L13)
* [supports/focus-img-ismap.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-img-ismap.js#L14)
* [supports/focus-img-usemap-tabindex.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-img-usemap-tabindex.js#L14)
* [supports/focus-redirect-img-usemap.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-redirect-img-usemap.js#L12)

### media-src: data

covered by directive `media-src data:;`

* [supports/focus-audio-without-controls.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-audio-without-controls.js#L12)
* [supports/focus-video-without-controls.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-video-without-controls.js#L12)

### object-src: data

covered by directive `object-src 'self' data:;`

* [query/focusable.strict.js](https://github.com/medialize/ally.js/blob/1.1.0/src/query/focusable.strict.js#L12)
* [supports/detect-focus.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/detect-focus.js#L34)
* [util/node-array.js](https://github.com/medialize/ally.js/blob/1.1.0/src/util/node-array.js#L14)

above CSP violation reports don't make any sense to me. The reasons are probably:

* [supports/focus-object-svg-hidden.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-object-svg-hidden.js#L11)
* [supports/focus-object-svg.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-object-svg.js#L12)

### frame-src: data

covered by directive `frame-src data:;`

* [query/focusable.strict.js](https://github.com/medialize/ally.js/blob/1.1.0/src/query/focusable.strict.js#L12)
* [supports/detect-focus.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/detect-focus.js#L34)
* [util/node-array.js](https://github.com/medialize/ally.js/blob/1.1.0/src/util/node-array.js#L14)

above CSP violation reports don't make any sense to me. The reasons are probably:

* [supports/focus-in-hidden-iframe.js](https://github.com/medialize/ally.js/blob/1.1.0/src/supports/focus-in-hidden-iframe.js#L11)

---

We're not the first to face this issue. Angular introduced [ng-csp](https://docs.angularjs.org/api/ng/directive/ngCsp) to work this in lieu of a proper [CSP feature detection](https://mikewest.org/2012/05/content-security-policy-feature-detection).

I don't yet know how CSP compliance should be handled by libraries such as ally.js. I'm open to suggestions.

**Comment by bvancea** [comment]
@rodneyrehm I've also ran into some issue with CSP, mainly due to the `focus-object-svg.js` test. This causes issues with with  `Content Security Policy directive: "object-src 'none'"`. 

I am curious, is there a specific reason for using an 'object' tag instead of an 'img' tag in those tests?

**Comment by rodneyrehm** [comment]
> I am curious, is there a specific reason for using an 'object' tag instead of an 'img' tag in those tests?

yes. The test is evaluating if an `<object>` element can be given focus if it hosts an SVG document.
